fido/Crypto/Fido2/Model/Binary/Encoding.hs

```diff
@@ -0,0 +1,183 @@
+
+    flags :: Word8
+    flags =
+      userPresentFlag
+        .|. userVerifiedFlag
+        .|. attestedCredentialDataPresentFlag
+        .|. extensionsPresentFlag
+      where
+        userPresentFlag = if M.adfUserPresent adFlags then Bits.bit 0 else 0
+        userVerifiedFlag = if M.adfUserVerified adFlags then Bits.bit 2 else 0
+        attestedCredentialDataPresentFlag = case sing @t of
+          SCreate -> Bits.bit 6
+          SGet -> 0
+        extensionsPresentFlag = case adExtensions of
+          Just _ -> Bits.bit 7
+          Nothing -> 0
+
+    builder :: Builder
+    builder =
+      Binary.execPut (Binary.putByteString $ convert $ M.unRpIdHash adRpIdHash)
+        <> Binary.execPut (Binary.putWord8 flags)
+        <> Binary.execPut (Binary.putWord32be $ M.unSignatureCounter adSignCount)
+        <> ( case sing @t of
+               SCreate -> encodeAttestedCredentialData rawAttestedCredentialData
+               SGet -> mempty
+           )
+        <> maybe mempty encodeExtensions adExtensions
+
+encodeRawAttestedCredentialData :: forall t raw. SingI t => M.AttestedCredentialData t raw -> M.AttestedCredentialData t 'True
+encodeRawAttestedCredentialData = case sing @t of
+  SCreate -> \M.AttestedCredentialData {..} ->
+    M.AttestedCredentialData
+      { acdCredentialPublicKeyBytes =
+          M.WithRaw $ LBS.toStrict $ CBOR.toLazyByteString $ encodePublicKey acdCredentialPublicKey,
+        ..
+      }
+  SGet -> \M.NoAttestedCredentialData -> M.NoAttestedCredentialData
+
+encodeAttestedCredentialData :: M.AttestedCredentialData 'M.Create 'True -> Builder
+encodeAttestedCredentialData M.AttestedCredentialData {..} =
+  Binary.execPut (Binary.putByteString $ M.unAAGUID acdAaguid)
+    <> Binary.execPut (Binary.putWord16be credentialLength)
+    <> Binary.execPut (Binary.putByteString $ M.unCredentialId acdCredentialId)
+    <> Binary.execPut (Binary.putByteString $ M.unRaw acdCredentialPublicKeyBytes)
+  where
+    credentialLength :: Word16
+    credentialLength = fromIntegral $ BS.length $ M.unCredentialId acdCredentialId
+
+encodeExtensions :: M.AuthenticatorExtensionOutputs -> Builder
+encodeExtensions M.AuthenticatorExtensionOutputs {} = CBOR.toBuilder $ CBOR.encode entries
+  where
+    entries :: HashMap Text CBOR.Term
+    entries = HashMap.empty
+
+encodeRawCollectedClientData :: forall t raw. SingI t => M.CollectedClientData t raw -> M.CollectedClientData t 'True
+encodeRawCollectedClientData M.CollectedClientData {..} = M.CollectedClientData {..}
+  where
+    ccdRawData = M.WithRaw $ LBS.toStrict $ toLazyByteString builder
+
+    builder :: Builder
+    builder =
+      stringUtf8 "{\"type\":"
+        <> jsonBuilder typeValue
+        <> stringUtf8 ",\"challenge\":"
+        <> jsonBuilder challengeValue
+        <> stringUtf8 ",\"origin\":"
+        <> jsonBuilder originValue
+        <> stringUtf8 ",\"crossOrigin\":"
+        <> jsonBuilder crossOriginValue
+        <> stringUtf8 "}"
```

Comment:
We should probably refer to https://www.w3.org/TR/webauthn-2/#clientdatajson-serialization here as well. I can imagine the later reference to this part of the spec can be missed.